license-details/README.md

```diff
@@ -0,0 +1,147 @@
+allowed.
+
+
+# Source files that cannot be released under the Apache 2.0 license
+
+## Python source files that import the Scapy package
+
+Note: All of the source files in this section are _only_ for running
+tests, and are _not_ part of a compiled `p4c` executable.
+
+The Python library Scapy is released under the GNU General Public
+License v2.0 [2].
+
+[2] https://github.com/secdev/scapy/blob/master/LICENSE
+
+Note: Scapy is released under the GPL, _not_ the Lesser GPL (aka
+LGPL).
+
+While the LGPL explicitly allows releasing proprietary binaries that
+dynamically link with LGPL libraries, the GPL has no such provisions.
+
+Unless someone can provide a legal ruling from an intellectual
+property lawyer to the contrary, we believe we are required to release
+all Python source files meeting the following conditions under the GPL
+v2.0 license:
+
+(a) files that directly import the `scapy` or `ptf` packages (the
+    `ptf` package imports `scapy`).
+(b) files that indirectly import the Scapy package, by importing a
+    package that imports Scapy
+
+We believe that this should not be a significant hindrance to
+commercial users of this code, because all such Python source files
+are test programs, run either during CI for testing p4c executables,
+or during testing by P4 developers.  These Python source files that
+use Scapy are not part of P4 compiler executables.
+
+If a company wishes to distribute Python programs that import the
+Scapy package, they will have to decide how to comply with Scapy's
+license themselves.  P4.org can inform them of what we believe the
+software license of Scapy requires, and make all efforts to keep Scapy
+out of P4 compiler executables, or other programs that we expect that
+companies will want to add proprietary extensions to.
+
+
+## C files intended to be compiled and executed in the Linux kernel, e.g. EBPF
+
+Note: When analyzing the source files in this section more carefully,
+we will take special care to ensure that they are not files used to
+build a P4 compiler executable, and update this note when that task is
+complete with our findings.
+
+One or more published P4C back ends produce C source files that are
+intended to be compiled and loaded into the kernel via the EBPF
+feature, at least [3], [4] and perhaps others.
+
+[3] https://github.com/p4lang/p4c/blob/main/backends/ebpf/README.md
+[4] https://github.com/p4lang/p4c/blob/main/backends/tc/README.md
+
+While there are a few examples of companies releasing binaries of
+Linux driver code for their hardware and not releasing the source
+code, we leave it to their legal teams to navigate any of the
+intricacies that may be involved in that endeavor.
+
+Our plan is to release all of the following kinds of source files
+under the GPL v2.0 license, the same as the Linux kernel:
+
+(a) files that are intended to be compiled and loaded into the kernel
+    via EBPF.
+(b) header files included from files in category (a).
```

Comment:
Are these files directly using a GPL content? Or are they just intended to be used in GPL-licensed "product" and Apache is not compatible with that (in which case maybe we should dual-license them).

Reply:
Whether Apache 2.0 is or not compatible with GPL is an open debate, IANAL, but as far as I know, this was never tested in the court. One way to workaround it is to introduce an exception (like in LLVM).